File: ts/src/app/ignore-rules.ts
import type { IgnoreConfig, SpendingConfig, TransactionIgnoreRule } from '../config.js';

export type TransactionIgnoreInput = {
  account_id: string;
  account_name: string;
  connection_id: string;
  connection_name: string;
  synchronizer: string;
  description: string;
  status: string;
  amount: string;
};

const EMPTY_SPENDING_CONFIG: SpendingConfig = {
  ignore_accounts: [],
  ignore_connections: [],
  ignore_tags: [],
};

type CompiledTransactionIgnoreRule = {
  account_id?: RegExp;
  account_name?: RegExp;
  connection_id?: RegExp;
  connection_name?: RegExp;
  synchronizer?: RegExp;
  description?: RegExp;
  status?: RegExp;
  amount?: RegExp;
};

function compileField(
  ruleIndex: number,
  field: keyof CompiledTransactionIgnoreRule,
  pattern: string | undefined,
): RegExp | undefined {
  if (pattern === undefined) return undefined;
  const trimmed = pattern.trim();
  if (trimmed === '') return undefined;

  let source = trimmed;
  const flags = new Set<string>();
  while (true) {
    const match = source.match(/^\(\?([a-z]+)\)/);
    if (match === null) break;

    for (const flag of match[1]) {
      if (flag !== 'i' && flag !== 'm' && flag !== 's') {
        throw new Error(
          `Invalid ignore.transaction_rules[${String(ruleIndex)}].${String(field)} regex: ${trimmed} (unsupported inline flag '${flag}')`,
        );
      }
      flags.add(flag);
    }
    source = source.slice(match[0].length);
  }

  const sortedFlags = [...flags].sort().join('');
  try {
    return new RegExp(source, sortedFlags);
  } catch (error) {
    const msg = error instanceof Error ? error.message : String(error);
    throw new Error(`Invalid ignore.transaction_rules[${String(ruleIndex)}].${String(field)} regex: ${trimmed} (${msg})`);
  }
}

function compileRule(ruleIndex: number, rule: TransactionIgnoreRule): CompiledTransactionIgnoreRule {
  const compiled: CompiledTransactionIgnoreRule = {
    account_id: compileField(ruleIndex, 'account_id', rule.account_id),
    account_name: compileField(ruleIndex, 'account_name', rule.account_name),
    connection_id: compileField(ruleIndex, 'connection_id', rule.connection_id),
    connection_name: compileField(ruleIndex, 'connection_name', rule.connection_name),
    synchronizer: compileField(ruleIndex, 'synchronizer', rule.synchronizer),
    description: compileField(ruleIndex, 'description', rule.description),
    status: compileField(ruleIndex, 'status', rule.status),
    amount: compileField(ruleIndex, 'amount', rule.amount),
  };

  const hasAnyField = Object.values(compiled).some((v) => v !== undefined);
  if (!hasAnyField) {
    throw new Error(`ignore.transaction_rules[${String(ruleIndex)}] must specify at least one regex field`);
  }
  return compiled;
}

export function compileTransactionIgnoreRules(config: IgnoreConfig): CompiledTransactionIgnoreRule[] {
  return compileTransactionIgnoreRulesWithSpending(config, EMPTY_SPENDING_CONFIG);
}

function escapeRegexLiteral(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function exactCaseInsensitivePattern(raw: string): string | undefined {
  const trimmed = raw.trim();
  if (trimmed === '') return undefined;
  return `(?i)^${escapeRegexLiteral(trimmed)}$`;
}

function synthesizeSpendingIgnoreRules(spending: SpendingConfig): TransactionIgnoreRule[] {
  const derived: TransactionIgnoreRule[] = [];

  for (const value of spending.ignore_accounts) {
    const pattern = exactCaseInsensitivePattern(value);
    if (pattern === undefined) continue;
    derived.push({ account_id: pattern });
    derived.push({ account_name: pattern });
  }

  for (const value of spending.ignore_connections) {
    const pattern = exactCaseInsensitivePattern(value);
    if (pattern === undefined) continue;
    derived.push({ connection_id: pattern });
    derived.push({ connection_name: pattern });
  }

  return derived;
}

export function compileTransactionIgnoreRulesWithSpending(
  config: IgnoreConfig,
  spending: SpendingConfig,
): CompiledTransactionIgnoreRule[] {
  const combined: TransactionIgnoreRule[] = [
    ...config.transaction_rules,
    ...synthesizeSpendingIgnoreRules(spending),
  ];
  return combined.map((rule, idx) => compileRule(idx, rule));
}

function matchField(pattern: RegExp | undefined, value: string): boolean {
  return pattern ? pattern.test(value) : true;
}

export function shouldIgnoreTransaction(
  compiledRules: CompiledTransactionIgnoreRule[],
  input: TransactionIgnoreInput,
): boolean {
  return compiledRules.some(
    (rule) =>
      matchField(rule.account_id, input.account_id) &&
      matchField(rule.account_name, input.account_name) &&
      matchField(rule.connection_id, input.connection_id) &&
      matchField(rule.connection_name, input.connection_name) &&
      matchField(rule.synchronizer, input.synchronizer) &&
      matchField(rule.description, input.description) &&
      matchField(rule.status, input.status) &&
      matchField(rule.amount, input.amount),
  );
}

File: src/app/ignore_rules.rs
use anyhow::{Context, Result};
use regex::Regex;

use crate::config::{IgnoreConfig, TransactionIgnoreRule};

#[derive(Debug, Clone)]
pub struct TransactionIgnoreInput<'a> {
    pub account_id: &'a str,
    pub account_name: &'a str,
    pub connection_id: &'a str,
    pub connection_name: &'a str,
    pub synchronizer: &'a str,
    pub description: &'a str,
    pub status: &'a str,
    pub amount: &'a str,
}

#[derive(Debug, Clone)]
pub struct TransactionIgnoreMatcher {
    rules: Vec<CompiledTransactionIgnoreRule>,
}

impl TransactionIgnoreMatcher {
    pub fn from_config(config: &IgnoreConfig) -> Result<Self> {
        let mut compiled = Vec::with_capacity(config.transaction_rules.len());
        for (idx, rule) in config.transaction_rules.iter().enumerate() {
            compiled.push(CompiledTransactionIgnoreRule::new(idx, rule)?);
        }
        Ok(Self { rules: compiled })
    }

    pub fn is_match(&self, input: &TransactionIgnoreInput<'_>) -> bool {
        self.rules.iter().any(|rule| rule.is_match(input))
    }
}

#[derive(Debug, Clone)]
struct CompiledTransactionIgnoreRule {
    account_id: Option<Regex>,
    account_name: Option<Regex>,
    connection_id: Option<Regex>,
    connection_name: Option<Regex>,
    synchronizer: Option<Regex>,
    description: Option<Regex>,
    status: Option<Regex>,
    amount: Option<Regex>,
}

impl CompiledTransactionIgnoreRule {
    fn compile_field(index: usize, field: &str, value: &Option<String>) -> Result<Option<Regex>> {
        let Some(pattern) = value else {
            return Ok(None);
        };
        let trimmed = pattern.trim();
        if trimmed.is_empty() {
            return Ok(None);
        }
        let compiled = Regex::new(trimmed).with_context(|| {
            format!("Invalid ignore.transaction_rules[{index}].{field} regex: {trimmed}")
        })?;
        Ok(Some(compiled))
    }

    fn new(index: usize, rule: &TransactionIgnoreRule) -> Result<Self> {
        let compiled = Self {
            account_id: Self::compile_field(index, "account_id", &rule.account_id)?,
            account_name: Self::compile_field(index, "account_name", &rule.account_name)?,
            connection_id: Self::compile_field(index, "connection_id", &rule.connection_id)?,
            connection_name: Self::compile_field(index, "connection_name", &rule.connection_name)?,
            synchronizer: Self::compile_field(index, "synchronizer", &rule.synchronizer)?,
            description: Self::compile_field(index, "description", &rule.description)?,
            status: Self::compile_field(index, "status", &rule.status)?,
            amount: Self::compile_field(index, "amount", &rule.amount)?,
        };

        let has_any_field = compiled.account_id.is_some()
            || compiled.account_name.is_some()
            || compiled.connection_id.is_some()
            || compiled.connection_name.is_some()
            || compiled.synchronizer.is_some()
            || compiled.description.is_some()
            || compiled.status.is_some()
            || compiled.amount.is_some();

        if !has_any_field {
            anyhow::bail!(
                "ignore.transaction_rules[{index}] must specify at least one regex field"
            );
        }

        Ok(compiled)
    }

    fn match_field(pattern: &Option<Regex>, value: &str) -> bool {
        pattern
            .as_ref()
            .map(|re| re.is_match(value))
            .unwrap_or(true)
    }

    fn is_match(&self, input: &TransactionIgnoreInput<'_>) -> bool {
        Self::match_field(&self.account_id, input.account_id)
            && Self::match_field(&self.account_name, input.account_name)
            && Self::match_field(&self.connection_id, input.connection_id)
            && Self::match_field(&self.connection_name, input.connection_name)
            && Self::match_field(&self.synchronizer, input.synchronizer)
            && Self::match_field(&self.description, input.description)
            && Self::match_field(&self.status, input.status)
            && Self::match_field(&self.amount, input.amount)
    }
}
